pkg/apipanic/apipanic.go

```diff
@@ -0,0 +1,37 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apipanic
+
+import (
+	"net/http"
+
+	"github.com/cilium/cilium/pkg/logging"
+)
+
+// APIPanicHandler recovers from API panics and logs encountered panics
+type APIPanicHandler struct {
+	Next http.Handler
+}
+
+// ServeHTTP implements the http.Handler interface.
+// It recovers from panics of all next handlers and logs them
+func (h *APIPanicHandler) ServeHTTP(r http.ResponseWriter, req *http.Request) {
+	defer func() {
+		if r := recover(); r != nil {
+			logging.DefaultLogger.WithField("panic_message", r).Warn("Cilium API handler panicked")
```

Comment:
I wonder if we can use WithError(r) since it's basically an error string.
Another possibility for others to comment on is that we only do this for the one broken pipe error.

Reply:
oh, also, the request object is in scope here. It might be good to print the path and client's IP to help debug this later. I'm not sure if there's anything else in there that would be handy to have (perhaps there's a time since request start that we could also print).

Reply:
@raybejjani it's not guaranteed that the panic will have error type, that's why I used WithField here. Will try to log a bit more information here, it makes perferct sense.